add api-only doc (#3640)
jupyterhub/jupyterhub: docs/source/reference/api-only.md

(api-only)=

# Deploying JupyterHub in "API only mode"

> JupyterHub is a service for deploying and managing Jupyter servers for users.

Reviewer: ```suggestion
As a service for deploying and managing Jupyter servers for users, JupyterHub
```

> This functionality is exposed _primarily_ via a [REST API](rest).

Reviewer: ```suggestion
exposes this functionality _primarily_ via a [REST API](rest).
```

> JupyterHub also ships with a _basic_ web UI built on that API,

Reviewer: ```suggestion
For convenience, JupyterHub also ships with a _basic_ web UI built using that REST API.
```

> enabling users to click a button to quickly start and stop their servers,

Reviewer: ```suggestion
The basic web UI enables users to click a button to quickly start and stop their servers,
```

> and admins to perform some basic user and server management tasks.

Reviewer: ```suggestion
and it lets admins perform some basic user and server management tasks.
```

> The REST API has always provided functionality not available in the UI,

Reviewer: ```suggestion
The REST API has always provided additional functionality beyond what is available in the basic web UI.
```

> and we try to avoid having any functionality only available in the UI and not the API.

Reviewer: ```suggestion
Similarly, we avoid implementing UI functionality that is also not available via the API.
```

> With JupyterHub 2.0, this should be fully the case -

Reviewer: ```suggestion
With JupyterHub 2.0, the basic web UI will **always** be composed using the REST API.
```

> no UI pages should rely on information not available via the REST API.

Reviewer: ```suggestion
In other words, no UI pages should rely on information not available via the REST API.
```

Previously, some admin UI functionality could only be achieved via admin pages,

> such as paginated requests.

Reviewer: With JupyterHub 2.0, web pages will no longer provide functionality that is unavailable by API.

> The JupyterHub UI is customizable via extensible HTML [templates](templates),

Reviewer: ```suggestion
## Limited UI customization via templates
The JupyterHub UI is customizable via extensible HTML [templates](templates),
```

but this has some limited scope to what can be customized.
Adding some content and messages to existing pages is well supported,
but changing the page flow and what pages are available are beyond the scope of what is customizable.

> Increasingly, JupyterHub is used purely as an API for managing Jupyter servers

Reviewer: ```suggestion
## Rich UI customization with REST API based apps
Increasingly, JupyterHub is used purely as an API for managing Jupyter servers
```
